pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/OwnershipCache.java

```diff
@@ -170,7 +170,17 @@ public CompletableFuture<Optional<NamespaceEphemeralData>> getOwnerAsync(Namespa
 
         // If we're not the owner, we need to check if anybody else is
         String path = ServiceUnitUtils.path(suName);
-        return lockManager.readLock(path);
+        return lockManager.readLock(path).thenCompose(owner -> {
+            //If the current broker is the owner, we must try to acquire ownership to avoid cache loss.
```

Comment:
What will happen if it's removed from the cache? I don't get the benefit here

Reply:
The client to loop through reconnecting to the current broker.

Reply:
Is it possible to reproduce this in a test? Why does it get into a loop?

Reply:
I have the same question. Could you explain more?

Reply:
“org.apache.pulsar.client.api.BrokerServiceLookupTest#testLookupConnectionNotCloseIfGetUnloadingExOrMetadataExalways” fail locally.

Could you try to run this test?

The root cause is the topic’s ownership not found in the ownedBundlesCache, and then the broker sends an error to the client, so you see the client is reconnecting. When the ownership is never found, the client keeps trying reconnect. Please check the WARN log from the PR description for details. I hope this log can help you.

Reply:
@nodece 

I can not find a test named `testLookupConnectionNotCloseIfGetUnloadingExOrMetadataExalways` in the test class named `org.apache.pulsar.client.api.BrokerServiceLookupTest`

Reply:
@poorbarcode `testLookupConnectionNotCloseIfGetUnloadingExOrMetadataEx` is correct.

Reply:
Seems like we are adding a caching logic in this getter func. Seems like we are doing more work here although the original intention of this func is to simply fetch the lock info from the metadata store.


I think we need to confirm 

1. if there is any side effect from adding this logic
2. Change the func name and doc that clarifies it also fills the cache.


Usually, we fetches any metadata info from the cache which refreshes and does async load if not already cached. I am wondering why we are not following this practice here for this ownership check, or we missed anything that the original author intended here.

Reply:
Can we explain when/how this ownership cache and metadata store are out of sync?

Reply:
Seems like we don't invalidate ownedBundlesCache when deleting the ownership. (Maybe I missed something here)

https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/OwnershipCache.java#L223-L233

It appears that we probably need to make those 

locallyAcquiredLocks
ownedBundlesCache
metadataStore

synchronized better.

Reply:
I guess we might see a racing condition here, when the same lock has been re-acquired immediately after release.

- locallyAcquiredLocks.put(namespaceBundle, rl) (new lock creation)
- invalidateLocalOwnerCache(namespaceBundle) (by lock.release deletes the ownedBundlesCache later)

ref: https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/OwnershipCache.java#L100-L107

Reply:
This PR looks ok to fix the possible cache and metadata inconsistency state.

I think we can also try to skip `expiredFuture.complete(null);` if the lock is recreated on the same broker.

However, I think we better make this ownership release protocol clearer not to cause such inconsistency states. 
In ExtensibleLoadBalancer, we release the topic resources earlier in the first state(Releasing state) and blocks(ignores by conflict resolver) further re-assignment until the current transfer protocol finishes. So, I think we should see less such racing conditions in ExtensibleLoadBalancer.

Reply:
@heesung-sn Thanks for your explanation. I think the ownership release protocol may work fine.

> This PR looks ok to fix the possible cache and metadata inconsistency state.

All right. I'm not sure if the issue is only in the pulsar testing environment.

Please notice that the `org.apache.pulsar.client.api.BrokerServiceLookupTest.BundleOfTopic` class in the test.

```
        private void releaseBundleLockAndMakeAcquireFail() throws Exception {
            ownedBundlesCache.synchronous().invalidateAll();
            mockZooKeeper.delete(ServiceUnitUtils.path(namespaceBundle), -1);
            mockZooKeeper.setAlwaysFail(KeeperException.Code.OPERATIONTIMEOUT);
        }
```

This method directly discards the cache and deletes zk data,  which may break the test case.

I will double-check the ownership mechanism to avoid the racing condition.